python architecture documentation (#132)
exasim-project/FoamAdapter: doc/development/architecture.rst

Architecture
============

FoamAdapter is a multi-language repository containing both C++ and Python code.
This document describes the architecture of FoamAdapter, including its C++ core and Python interface components.

.. note::
   This section of the documentation provides:
     * a high-level overview of the *planned* architecture 

>      * guidance through the review process

Reviewer: ```suggestion
```

Author: Will be removed with stack PR

     * example implementations serving only as proof of concept to illustrate the *planned* architecture
     * a note that detailed features will evolve and refine the architecture, updating examples as development progresses

Overview
--------

> FoamAdapter is designed as a multi-physics, Python-based simulation framework.

Reviewer: ```suggestion
NeoFOAM is designed as a multi-physics, Python-based simulation framework.
```

Additionally, it includes a set of C++-based legacy solvers, such as neoIcoFoam, to simplify the transition from existing OpenFOAM solvers and workflows.
It provides a flexible and modular architecture that allows users to easily extend and customize their simulation setup.

The architecture provides the following features to achieve the goals outlined in the :doc:`goals and features <../goals_features>` document:

- Easy coupling of multiple domains and physical models
- Field and model initialization based on dependency graphs
- Modular solver design that computes data dependencies at runtime
- Plugin architecture for extending models and fields

> To support multi-physics capabilities, multiple computational domains are implemented.  

Reviewer: ```suggestion
To support multi-physics capabilities, multiple computational domains are supported.  
```

Each domain has **one solver** assigned, which defines the governing equations, operations, and **optional additional physical models.**

> Multiple domains can be defined via input files, and coupling between domains is automatically handled based on the selected physics modules.

Reviewer: ```suggestion
Coupling between domains is automatically handled based on the selected physics modules.
```

Reviewer: I dont understand the part about the files, also i think that is too specific for a feature which is not there yet.

            STEP1["Solver </br> Momentum Equation"]
            STEP2["Added by Model </br> Temperature Equation"]
            STEP3["Solver </br> Continuity Equation"]
            STEP4["Solver </br> Update Turbulence"]
        end
        
        STEP1 --> STEP2
        STEP2 --> STEP3
        STEP3 --> STEP4
        
        %% Physics Extensions (simplified)
        subgraph AddPhysics ["Additional Physics Modules"]
            direction TB
            POROSITY["Porosity"]
            ROTATION["Rotating Reference Frame"]
            BUOYANCY["Boussinesq Approximation"]
        end
        POROSITY -.-> STEP1
        ROTATION -.-> STEP1
        BUOYANCY -.-> STEP2
        BUOYANCY -.-> STEP3

        style MAIN fill:#E3F2FD
        style AddPhysics fill:#E3F2FD
        style STEP1 fill:#2196F3,color:#fff
        style STEP2 fill:#FF9800,color:#fff
        style STEP3 fill:#9C27B0,color:#fff
        style STEP4 fill:#607D8B,color:#fff

The solver defines the main operations to be executed: momentum, continuity, and turbulence model updates.
Additional physics models, such as porosity, rotation, or buoyancy, can modify or add operations.

> This relationship is illustrated in the pseudocode below:

Reviewer: Somehow this is not rendered here https://exasim-project.com/NeoFOAM/Build_PR_132/development/architecture.html

Reviewer: Also I still don't fully understand the pseudo code example. Maybe because porosity and rotation are missing. How does BoussinegsModel interact with IncompressibleFluidSolver? Maybe this is not fully done yet?

Author: section now reads

```
The solver defines the main operations to be executed: momentum, continuity, and turbulence model updates.
Additional physics models, such as porosity, rotation, or buoyancy, can modify or add operations.
In this case, the porosity model and the Rotating Reference Frame add momentum source terms, while the Boussinesq model adds a temperature equation and modifies the continuity equation.

Each solver maintains a list of additional physics models that can add or modify operations.
These additional physics models are defined externally from the solver and only need to comply with the IncompressibleFluidModel interface, which can be defined separately by each solver.
The execution order is determined at runtime based on the metadata specified in each model.
```

.. code-block:: python
    # Pseudocode illustrating solver and model structure
    @Solver
    class IncompressibleFluidSolver:

>         models: list[IncompressibleFluidModel]  # Additional physics models

Reviewer: I dont understand the relation between IncompressibleFluidModel and IncompressibleFluidSolver?

Author: see above

Architecture
============

> This document describes the overall architecture of FoamAdapter, including both the C++ core and Python interface components.

Reviewer: ```suggestion
FoamAdapter is multilanguage repository as it contains both C++ and python code.
This document describes the architecture of FoamAdapter, including the C++ core and Python interface components.
```

Overview
--------

> FoamAdapter is designed as a multi-physics, python-based simulation framework.

Reviewer: ```suggestion
FoamAdapter is designed as a multi-physics, python-based simulation framework.
Additionally, a set of C++ based legacy solvers like neoIcoFoam simplify the transition from existing OpenFOAM solvers and workflows.
```

It provides a flexible and modular architecture that allows users to easily extend and customize the simulation setup.

The architecture provides following features to achieve the goals outlined in the :doc:`goals and features <../goals_features>` document:

- Easy coupling of multiple domains and physics
- Field and model initialization based on dependency graphs
- Modular solver design that computes the data dependencies at runtime
- Plugin architecture for easy extension with new models and fields

> The main abstraction is that each domain has one physics module assigned to it that defines the governing equations and models for that domain.

Reviewer: ```suggestion
In order to implement multi-physics capabilities multliple computational domains are supported.  
Each computational domain has one physics module assigned that defines the governing equations and physical models.
```

Multiple domains can be defined with input files and the coupling between the domains is handled automatically based on the defined physics modules.

> The following sections describe the main feature and and implementation example to give a high level overview of the architecture.

Reviewer 2: ```suggestion
The following sections describe the main feature and implementation example to give a high level overview of the architecture.
```
`and` is repeated.

> More details are described in the section for each feature. 

Reviewer: ```suggestion
```

Solver Execution Model
----------------------

FoamAdapter defines solver behavior as a dependency-driven execution model.

> Each solver declares its required fields and models, and the runtime system constructs a directed acyclic graph (DAG) to determine the correct execution order.

Reviewer: So does that mean that each computational domain has one solver attached? I guess that should be mentioned above

> Each solver declares its required fields and models, and the runtime system constructs a directed acyclic graph (DAG) to determine the correct execution order.

Reviewer: ```suggestion
Each solver declares required fields, e.g. pressure, temperature, etc and physical models. 
The ?? then constructs a directed acyclic graph (DAG) to determine the order in which fields are solved for and the models need to be executed.
```

The following diagram illustrates the workflow for two common solver types for a conjugate heat transfer scenario:

.. mermaid::

    %%{init: {'flowchart': { 'htmlLabels': true, 'wrap': true }}}%%
    flowchart TB

>         subgraph setup1 ["setup "]

Reviewer: Maybe "setup domain 1" and "setup domain 2" or solid and fluid ...

            direction TB

>             S1A["Initialize Fields"]

Reviewer: Which fields?

        end

        subgraph setup2 ["setup"]
            direction TB
            S2A["Initialize Temperature Field"]
        end

>         subgraph S1 ["non-thermal-fluid solver"]

Reviewer: I think it is easier to understand with just fluid solver and solid solver.

            direction TB
            S1C["Momentum Predictor"]
            S1C --> S1E["Solve Energy Equation"]
            S1E --> S1F["Pressure Corrector<br/>PISO Loop"]
            S1F --> S1J["Update turbulence Model"]
        end

        subgraph S2 ["thermal-solid solver"]
            direction TB
            S2B["Solve Energy Equation"] --> S2D["Update Solid Properties"]
        end

        setup1 --> S1C
        setup2 --> S2B
        S1E --> S2B
        S2B --> S1E

The diagram shows two selected solvers: a non-thermal-fluid solver and a thermal-solid solver.
Each solver has its own setup phase and computational steps, with interactions between the two solvers for energy exchange.

> The solution steps of each solver provide the dependencies and the runtime system constructs a DAG to determine the correct execution order based on these dependencies.

Reviewer: I think that is basically repeated from before
```suggestion
```
I think instead it would be good to show the DAG here and refere to it.

> .. code-block:: python

Reviewer: Is that user facing code, or internal? I think we need a bit text here discussing how the python code correlates with the "flow chart" above.

    # 'model' is a placeholder for the model dependency object required by each step.
    model = ...  # placeholder for model dependency

>     @model("fluidSolver")

Reviewer: Needs a sentence explaining the decorator.

    class DomainA:

>         @step(order=1, model)

Reviewer: Does it really need `model` here? Maybe to be a real DAG `depends_on`  would be required.

>         def momentum(self, ctrl): pass

Reviewer: what is `ctrl`?

        @step(order=2, model)
        def solve_energy(self, ctrl): pass
        @step(order=3, model)
        def pressure_corrector(self, ctrl): pass
        @step(order=4, model)
        def update_turbulence(self, ctrl): pass

    @model("solidSolver")
    class DomainB:
        @step(order=1, model)
        def solve_energy(self, ctrl): pass
        @step(order=2, model)
        def update_solid_properties(self, ctrl): pass

A solver is defined as a class with multiple steps, each step has to declare its dependencies.
Additional models can be executed before and after each step to modify the behavior and add additional physics.

This allows to easily extend existing solvers with new physics without modifying the core solver implementation and promotes code reuse.

.. note::

    The implementation is still work in progress but only details may change.

> Modular Solver Architecture

Reviewer: I have trouble understanding this section, what is the relation to the previous section? Can the previous example be extended, for example by mentioning that the fluid solver of the previous example can be extended by adding additional physics modules.

---------------------------

> The modular solver architecture briefly touched on in the previous section is a key feature of FoamAdapter and allows for flexible composition of complex multi-physics simulations.

Reviewer: Feels repetetive.
```suggestion
```

> An example of the modular solver architecture is illustrated in the following diagram:

Reviewer: My understanding is

```suggestion
The modular solver architectures allows to extend individiual solver by adding addititional physics modules.
This is illustrated in the following diagram, where the fluid solver is extended by three physics submodules.
```

Then proceed to discuss how and why individual modules interact with the other components.

            ROTATION["Rotating Reference Frame"]
            BUOYANCY["Boussinesq Approximation"]
        end
        POROSITY -.-> STEP1
        ROTATION -.-> STEP1
        BUOYANCY -.-> STEP2
        BUOYANCY -.-> STEP3

        style MAIN fill:#E3F2FD
        style AddPhysics fill:#E3F2FD
        style STEP1 fill:#2196F3,color:#fff
        style STEP2 fill:#FF9800,color:#fff
        style STEP3 fill:#9C27B0,color:#fff
        style STEP4 fill:#607D8B,color:#fff

In this architecture, the main solver loop consists of core steps such as momentum prediction and pressure correction.
Additional physics modules (e.g., porosity, rotation, buoyancy) can be plugged into the workflow to modify the behavior of these core steps.

This modular design enables users to easily add or remove physics effects without altering the fundamental solver structure, promoting code reuse and maintainability.
The additional physics modules need to define their dependencies and the runtime system ensures that they are executed at the correct point in the solver loop.

.. note::

    implementation is still work in progress

Field and Model Initialization
------------------------------

> As the fields and solver needs to be initialized in the correct order before the solver run, FoamAdapter provides a structured initialization phase to ensure that the necessary fields and models are properly set up.

Reviewer: I think this needs a bit of clarification what field initialization means, and how the order of initilization matters. If all fields are just read from disk there is no initialization order issue, if i understand correctly this refers to fields which are derived from other fields. One example could be the viscosity which might or might not depend on temperature depending on the physical model.

> The fields and models are stored lazily at first with the dependencies and a DAG is solved to determine the correct initialization order.

Reviewer: I don't understand the meaning of this sentence. I would say they are not **stored** lazily but initialized?

> .. image:: pimpleDag.png

Reviewer: The figure needs a bit of discussion. What do i see here, what does it tell me?

   :alt: DAG of field and model dependencies for pimpleFoam
   :align: center

> The following code snippet illustrates how a turbulence model and a derived field (face velocity) are defined with their dependencies:

Reviewer: Before discussing these details i think the basic building blocks like model, solver, module and the representation in python needs to be introducted, e.g. the decorators `@fields.Fields.deps` and others

            singlePhaseTransportModel = models.get_model(deps, "singlePhaseTransportModel")
            turbulence = incompressibleTurbulenceModel.New(U, phi, singlePhaseTransportModel)
            return turbulence

    @fields.Fields.deps("U")
    def create_face_velocity(deps: dict) -> surfaceScalarField:
        U = fields.get_field(deps, "U")
        phi_field = createPhi(U)
        return surfaceScalarField(
            value=phi_field,
            dimensions=(0, 1, -1, 0, 0, 0, 0),
            description="Face flux field",
        )

The same approach is used for all fields and models in FoamAdapter, ensuring a consistent and reliable initialization process.

Plugin Architecture
-------------------

Motivation
^^^^^^^^^^

Modern scientific and engineering workflows require flexible simulation frameworks that can be easily extended and customized.
FoamAdapter's plugin architecture is designed to enable users and developers to add new physics models, boundary conditions, and solver modules without modifying the core codebase.
This approach promotes maintainability, collaboration, and rapid prototyping of new features.

Concept
^^^^^^^

> FoamAdapter implements a runtime-extensible plugin/config system using Pydantic discriminated unions and a registry pattern.

Reviewer: Maybe make `Pydantic discriminated unions` a link to their documentation also for registry pattern.

The core idea is to allow new plugin types (e.g., models, fields, solvers) to be registered dynamically, either at runtime or via Python entry points (setuptools).
Each plugin type (such as physics models or boundary conditions) is managed by a registry, which collects all available plugin classes and exposes a unified configuration model for input validation and schema generation.

**Background: Pydantic Discriminated Unions**

> Pydantic supports discriminated unions for type-safe configuration, but the set of types in the union must be known at model definition time. For example:

Reviewer: ```suggestion
Pydantic supports discriminated unions for type-safe configuration, but the set of types in the union must be known at model definition time. 
The following example ensures that a pet is either Cat, Dog, or Lizard
```
Describe what else happens, i don't think it is very obvious

        style STEP3 fill:#9C27B0,color:#fff
        style STEP4 fill:#607D8B,color:#fff

The solver defines the main operations to be executed: momentum, continuity, and turbulence model updates.
Additional physics models, such as porosity, rotation, or buoyancy, can modify or add operations.
In this case, the porosity model and the Rotating Reference Frame add momentum source terms, while the Boussinesq model adds a temperature equation and modifies the continuity equation.

Each solver maintains a list of additional physics models that can add or modify operations.
These additional physics models are defined externally from the solver and only need to comply with the IncompressibleFluidModel interface, which can be defined separately by each solver.
The execution order is determined at runtime based on the metadata specified in each model.

.. code-block:: python

    # Pseudocode illustrating solver and model structure
    @Solver
    class IncompressibleFluidSolver:
        models: list[IncompressibleFluidModel]  # Additional physics models
        @Solver.step(...)
        def momentum(self, ...): pass
        @Solver.step(...)
        def continuity(self, ..

>         .): pass

Reviewer: ```suggestion
        def continuity(self, ...): pass
```
